### PR TITLE
Develop and release mregions v1.0.0

### PR DESCRIPTION
**Pull request by salvafern**
# Introduction

The [Marine Regions Gazetteer](https://marineregions.org/gazetteer.php) is a standard list of marine georeferenced place names.

> Gazetteer: a dictionary of geographical names. 

<sup>https://www.thefreedictionary.com/gazetteer</sup>

All geographic objects of the Marine Regions Gazetteer database have a unique ID, called the [**MRGID**](https://marineregions.org/mrgid.php)

> Syntax `http://marineregions.org/mrgid/<number>`

<sup>E.g. https://marineregions.org/mrgid/14</sup>

In addition to the Marine Regions Gazetteer, the Marine Regions Team (currently: @LennertSchepers , @brittlnv, @salvafern and @bart-v) creates and hosts data products, being the most popular one the [Marine Regions Maritime Boundaries](https://marineregions.org/eez.php).

140389_marineregions-v11---eez

<sup>(By @brittlnv. Source: https://www.vliz.be/en/news?p=show&id=8160)</sup>

These data products can be accessed via direct download, but also by using OGC WFS web services. They are hosted at the VLIZ geoserver: 
> `http://geo.vliz.be/geoserver/MarineRegions/wfs?service=WFS&request=GetCapabilities`

# What's new?

During 2021, new webservices were developed to access the Marine Regions Gazetteer. In essence, two major changes happened:
* The database can be accessed now through Open Linked Data standards: JSON-LD and TTL. 
* The geometry associated to each gazetteer record is now available through RESTful services.

More information in: 

> Lonneville B. et al. (2021) Publishing the Marine Regions Gazetteer as a Linked
> Data Event Stream. S4BioDiv 2021. http://ceur-ws.org/Vol-2969/paper8-s4biodiv.pdf

We want to update the mregions package to use these two new functionalities. 

# How?

**Gazetteer webservices**

We noticed that not all the gazetteer webservices were being used in the package. We want to add all them now, specifically:

* [ ] getGazetteerRecordByMRGID
* [ ] getGazetteerGeometry
* [ ] getGazetteerGeometries
* [ ] getGazetteerTypes
* [ ] getGazetteerRecordsByName
* [ ] getGazetteerRecordsByNames
* [ ] getGazetteerRecordsByType
* [ ] getGazetteerRecordsBySource
* [ ] getGazetteerRelationsByMRGID
* [ ] getGazetteerWMSes
* [ ] getGazetteerRecordsByLatLong
* [ ] getGazetteerSources
* [ ] getGazetteerNamesByMRGID

The list of the Gazetteer webservices is in: https://www.marineregions.org/gazetteer.php?p=webservices

**Functions naming**

We think that the names of the functions should reflect very clearly what is the service you are using to avoid any confusion. For instance: `mr_geo_code()` uses `getGazetteerRecordsByName`. We will change this name to something more similar like: `mr_gaz_records_by_name()`.

**RESTful formats**

You can get gazetteer records in four formats: json, xml, json-ld and turtle. We will get json-ld for the geometries and `getGazetteerRecordByMRGID`, and plain json for the rest of the gazetteer webservices. 

**R packages to be used**

To access the Gazetteer webservices, we will use `httr` . To parse json, we will use `jsonlite`. We have to look for packages that deal with json-ld. 

To access WFS, we will use `ows4R` .

**R data types**

The output from the functions in R either `tibbles` (no geometry) or `sf` objects (with geometry). Depending on the function it might make sense to use lists instead.

# To do:

**Gazetteer webservices**
* [ ] Create new functions to access all the gazetteer webservices
* [ ] Write automated tests with `testthat`
* [ ] Write documentation: Get Started vignette and README

**OGC webservices**
* [ ] Create new functions to access the Marine Regions data products via WFS
* [ ] Write automated tests with `testthat`
* [ ] Write documentation: Get Started vignette and README

**Other**
* [ ] Deprecate older functions
* [ ] Write advance use vignette
* [ ] Create helper functions (e.g. eezmapper etc)

### CODE REVIEW

**Comment by maelle** [comment]
Naive question: after all these changes, what will be left of the current package content?

**Comment by salvafern** [comment]
Good question :sweat_smile: not much. We expect to keep using the way of reading the Gazetteer RESTful API via `jsonlite` . 

But we would like to use functions names that are consistent with the names of the webservices. Plus we want to make a crystal clear distinction between (1) the Gazetteer database and (2) the maritime boundaries products, even in the names of the functions (e.g. `mr_gaz_*()` and `mr_ogc_*()`, we haven't decided the best naming yet) . Both things are part of Marine Regions but they work in different ways and must be accessed differently.

In other words, it will most likely look like a new package, although we will keep the previous functions as deprecated for reproducibility.

**Comment by maelle** [comment]
Right, because I was thinking that if it actually were a brand-new package e.g. mregions2 there would be no question that it'd be candidate for peer-review. :grin:

**Comment by noamross** [comment]
We have done re-reviews once or twice in the past (see https://github.com/ropensci/software-review/issues/29) and we think it is something to build in to the future of peer review - at least recording the package version and standards versions at the time of initial and each subsequent review.  While not required to maintain your badge, I think getting new reviews for this major overhaul makes and and we'll do our best to move the process along.

**Comment by salvafern** [comment]
Ok great! We will work on the package and submit for review once it is ready. Thanks for the help @maelle @noamross !

**Comment by marc-portier** [comment]
Some late suggestions and reactions:
(more from my focus and experience with the  LD parts of the MR services update of 2021)


## rdflib

most languages have a library to handle RDF and its many serializations (mainly in this case jsonld and turtle) 

luckily R is no exception --> https://cran.r-project.org/web/packages/rdflib/index.html

I have no practical experience with it, but it would be natural first candidate.  Maybe smart to have a selection of MRGIDs (in ranging nature of complexity) to make a collection of files (different formats) to test the parsing and handling + fit of using this lib


Might make sense to take into account the ontologies we chosen:
* Marine Regions gazetteer ontology documentation https://www.marineregions.org/ns/ontology
* Placetype ontology documentation https://www.marineregions.org/ns/placetypes (Turtle format only)
Together 

## httr usage & conneg

Since the MR service now supports content-negotiation I strongly advise using the accept header to retrieve the various formats.

```R
GET("https://marineregions.org/mrgid/26567", accept("text/turtle"))
GET("https://marineregions.org/mrgid/26567", accept("application/ld+json"))
```



## consuming LDES 

A new addition to the service offering is our [change-feed](https://www.marineregions.org/feed) in [LDES](https://w3id.org/ldes/specification) format. It offers a fragmentation of the whole dataset of MRGIDs based on `lastmodification` time.  Its primary use is to make it easy to have external caches, replications, or derived indexes or full-set-post-calculations update themselves in an efficient and incremental way.

Exposing that new service to R package users suggests we introduce a new `mr_updated_since(sincedate)` function.

The implementation of which would be to fetch the starting page at https://www.marineregions.org/feed and then follow the tree:relation paths down to pages that contain only tree:member entries that are older then the requested date.


## considering aggregating many geometries into one

Note that one MRGID can have multiple geometries (all of them being possible off MULTIPOLYGON type) that represent it.  These representations could be composing parts as well as competing alternative variants.  Since the distinction can not be made the general consumer advise is to either run over all of them or aggregate them all into one UNION and use that as the single all-encompassing view of the region.

There seem to be a number of existing strategies in R (https://gis.stackexchange.com/questions/278818/fastest-way-to-union-a-set-of-polygons-in-r) to create such a union.   

IMHO I think it is a useful addition to have this package do this for its users, rather then leave it as an exercise to them.


## in memory Object representation?

Maybe more then we are intending at this time, but (again as a higher-level service to the users) it could make sense to have actual R-Object-Class wrappers to represent each MRGID in an object-oriented and more user-friendly / semantic way.

Such more advanced layer should be able to further lower the threshold (hiding technical representation details like ttl, json, ...)  and thus open up possible use of this library.  Of course it can also be considered interesting future work, or better fitting into a separate package building on this one.

**Comment by maelle** [comment]
By the way maybe the rewrite is an opportunity to switch to httr2? https://httr2.r-lib.org/

**Comment by salvafern** [comment]
Thanks @marc-portier for all the help! We will definitely use your advice.

We have finally decided to create a second package, mregions2. This way we don't break backwards compatibility. We will request an rOpenSci review later when the package is developed.

Follow up in: https://github.com/lifewatch/mregions2/issues/1
